Gateways article - draft 2 (#2)
2color/ipfs-articles: gateways/break-on-through-gateways.md

- _path_ to the location on the server: `/sites/default/files/thumbnails/image/04_iss067e033423.jpg`

![location addressing](./location-addressing.png)

The challenges with location addressing are numerous. We've all had the experience of going down an internet rabbit hole only to be abrupted by dead links because the link changed or the server is no longer hosting the files.

In a peer-to-peer network like IPFS, a given file might be hosted on a number of the IPFS nodes.

This is where _content addressing_ comes in handy. With IPFS, every single file stored in the system is addressed by a [cryptographic hash](https://en.wikipedia.org/wiki/Cryptographic_hash_function) of its contents known as a [**Content Identifier** or **CID**](https://docs.ipfs.io/concepts/glossary/#cid). The CID is a long string of letters and numbers that is unique to that file.

There are three crucial things to remember with regards to CIDs:

- Any difference –even a single bit– to the file (or file tree in the case of a folder) will produce a different CID. This property is known as **immutability**.
- The same content added to two different IPFS nodes will produce the same CID (given the same parameters).

> - A single CID can represent a single file or a folder of files, e.g. a static website. This property is known as "turtles all the way down".

Reviewer: Just in case reader is bit dense ;-))

```suggestion
- A single CID can represent a single file or a folder of files, e.g. a static website. This property is known as "[turtles all the way down](https://en.wikipedia.org/wiki/Turtles_all_the_way_down)".
```

> ![Content addressing](./content-addressing.png)

Reviewer: Sidenote: in case you want to shorten a CID on diagram or in text, the convention is to show at least four characters from each side separated by ellipsis –  for example `bafy...hmfm`

Author: Great! Wasn't aware of this but I'll certainly update the diagram.

- Running an IPFS node by installing one of the IPFS implementations as a daemon (long-running process) on your computer or on a server in the cloud. The node becomes a member of the IPFS peer-to-peer network and announces what data it’s holding and responds to requests for data.
- Using an **IPFS Gateway** which allows fetching CIDs using the HTTP protocol.

The first option allows you to _speak the native IPFS protocol_ while the latter serves as a bridge in situations where you might be constrained to using HTTP. Choosing the right approach depends on your use case.

## What are IPFS gateways?

IPFS gateways are public services that translate between _Web2_ and _Web3_ thereby providing a bridge between HTTP and IPFS.

They allow you to use the HTTP protocol –which almost every programming language is capable of– to request a CID from the IPFS network, fetch it, and use HTTP to send the data back.

In its simplest form, a gateway is an IPFS node that also accepts HTTP requests for CIDs in addition to speaking the IPFS protocol to participate in the peer-to-peer network. Most IPFS implementations can also work as a gateway.

> ![Any IPFS node can also be a gateway](./browser-gateway.png)

Reviewer: nit: go-ipfs is only one of many implementations, some may not have gateway at all :)
```suggestion
![Some IPFS nodes can also be a gateway](./browser-gateway.png)
```

Author: Besides go-ipfs, are there any other ipfs implementations that can work as an HTTP gateway?

This means if your CID isn't well replicated in the IPFS network, fetching a CID from a gateway can take anywhere between milliseconds to the order of tens of seconds – obviously suboptimal if you're building an app that requires fast loading times.

In the next part, you will learn about some approaches to improve CID request latency and ensure reliable access to your CIDs when using IPFS gateways.

## Improving performance and reliability when fetching CIDs from IPFS

Given the decentralized nature of IPFS, improving the reliability and request latency of your CIDs through an IPFS gateway depends on many factors:

- Whether the IPFS gateway has the CID cached.
- Which gateway you are requesting the CID from?
- The amount of traffic and load on the IPFS gateway
- The number of IPFS nodes **pinning** the CID.
- Is there a [CDN](https://en.wikipedia.org/wiki/Content_delivery_network) in front of the IPFS gateway?
- The network distance (round-trip time) between the requester and the IPFS gateway.

> - HTTP Cache headers

Reviewer: ```suggestion
- [HTTP Cache](https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching) headers
```

Taking all these into consideration makes it hard to give generic advice. Nonetheless, understanding the factors influencing the performance can help you navigate the space of potential solutions.

Some are these factors are within your control. The section below will cover some concrete tips you can take to improve performance and reliability. But first, let's go into the subtle differences between pinning.

### Caching vs. pinning

As mentioned earlier, an IPFS gateway in its simplest form is an IPFS node with its HTTP gateway port open. When you request a CID, it will be returned quickly if the IPFS node has the CID cached or pinned. If not, it has to search the network.

> IPFS has a fairly aggressive caching mechanism that will keep an object local for a short time after the node has fetched it from the network, but these objects may get garbage-collected regularly.

Reviewer: nit: caching strategy is mostly up to implementation and specific deployment, perhaps somehting like this would be safer?

```suggestion
Existing IPFS [implementations](https://ipfs.io/#install) have a fairly aggressive caching mechanism that will keep an object local for a short time after the node has fetched it from the network, but these objects may get garbage-collected regularly, especially when a public gateway is used.
```

Author: Well spotted!

Pinning is the mechanism that allows you to tell IPFS to **always** store a given CID — by the default on your local node. In addition to local pinning, you can also pin your CIDs to [remote pinning services](https://docs.ipfs.io/how-to/work-with-pinning-services/).

In other words, caching is the mechanism by which CID is kept around on the node for a short period until garbage-collected while pinning is a deliberate choice you make to keep the CID stored on the node.

This is why requesting a CID for the first time from a gateway can take time while subsequent requests are typically faster.

### Tip 1: Pin your CIDs to multiple IPFS nodes

Drawing on the principles laid out above, it's sensible to pin your CIDs to multiple IPFS nodes to ensure reliable availability and fast fetching. These can be either IPFS nodes that you are operating or pinning services like [Web3.storage](https://web3.storage/), [Pinata](https://www.pinata.cloud/), and [Infura](https://infura.io/product/ipfs).

> To make pinning easier, there's a [Pinning Service API Spec](https://ipfs.github.io/pinning-services-api-spec/) that is already supported by IPFS node implementations, client libraries, and pinning services. Using this spec, you can implement pinning to multiple services as part of your workflow for uploading immutable data to IPFS.

Reviewer: ```suggestion
To make pinning easier, there's a vendor-agnostic [Pinning Service OpenAPI Specification](https://ipfs.github.io/pinning-services-api-spec/) that is [already supported by many IPFS node implementations, client libraries, and existing pinning services](https://github.com/ipfs/pinning-services-api-spec#adoption). Using this remote pinning API, you can [implement pinning to multiple services](https://docs.ipfs.io/how-to/work-with-pinning-services/#use-an-existing-pinning-service) as part of your workflow for uploading immutable data to IPFS.
```

> Note that pinning is not the same as adding the CID to the IPFS network. For pinning to work, the CID has to first be added to a reachable IPFS node connected to the network so that the pinning services can replicate the CID. All of the pinning services also offer an API for uploading a file that returns the CID.

Reviewer: ```suggestion
Note that pinning is not the same as adding the CID to the IPFS network. For pinning to work, the CID has to first be added to a reachable IPFS node connected to the network so that the pinning services can replicate the CID. As a convenience, most of the pinning services also offer an API for uploading a file that returns the CID.
```

If you're not running an IPFS node, you can start by uploading a file to one service and then using the returned CID to pin it to other services.

### Tip 2: Use a custom domain that you control as your IPFS gateway

Imagine the following scenario: you deploy your web app to IPFS which contains media with absolute URLs to a public gateway experiencing an outage. For example, your web app displays the image with an absolute path to the CID: https://bafybeibml5uieyxa5tufngvg7fgwbkwvlsuntwbxgtskoqynbt7wlchmfm.ipfs.infura-ipfs.io.

You may be able to reach your app using a different gateway, but since the web app's content is immutable, the image pointing to the Infura IPFS gateway which is down will not load.

> For this reason, it's sensible to use a domain within your control to route HTTP traffic to a gateway. This approach potentially gives you the flexibility to implement additional performance optimizations.

Reviewer: Running own gateway comes with additional control/safety around "hosting bad bits":

```suggestion
Also, you may also want to only host specific CIDs, such as data uploaded by your users, deny serving HTML websites, or block fetching any third party content in general.

For these reasons, it's sensible to use a domain within your control to route HTTP traffic to a gateway. This approach potentially gives you the flexibility to implement additional performance optimizations.
```

Practically speaking, this can be implemented using several approaches depending on your willingness to run infrastructure:

> - Point a domain you control, e.g., `*.ipfs.yourdomain.io` point to a reverse proxy like nginx which will proxy requests to a public gateway, allowing you to switch public gateways if there's downtime.

Reviewer 2: `*.ipfs.yourdomain.io` -> `*.ipns.yourdomain.io`

Reviewer: Probably need to list both: `*.ipfs.yourdomain.example.com` and `*.ipns.yourdomain.example.com`

- Use [Cloudflare workers](https://workers.cloudflare.com/) to implement a smart proxy that races a request across multiple gateways. This is the approach taken by the [NFT.storage gateway](https://nft.storage/docs/concepts/gateways/#architecture) for which you can find the [source code on GitHub](https://github.com/nftstorage/nftstorage.link/tree/main/packages/edge-gateway#high-level-architecture).

<!-- ### Tip 3: Make use of the  and etag headers -->

### Tips if you're running an IPFS node

- Explicitly peer with the pinning services that you use and public gateways
- Ensure that you are correctly returning HTTP cache headers to the client if the node is behind a reverse-proxy

> - Put a CDN like Cloudflare in front to reduce load on the IPFS node

Reviewer: Peering with pinning service makes sense only if you run your own gateway (you can't force a third-party public gateway to peer with someone you want), so maybe we should change the title of this section?

```suggestion
### Tips if you're running your own HTTP gateway

- Set up [peering](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#peering) with the pinning services that store your data
- Ensure that you are correctly returning HTTP cache headers to the client if the node is behind a reverse-proxy
- Put a CDN like Cloudflare in front to reduce load on the IPFS node
```

Example: NFT.storage has a new gateway that improves retrieval performance: nftstorage.link. This gateway “races” 3 public gateways (Pinata, Cloudflare, and ipfs.io) and also caches the majority of NFTs (>70% of them).
Individually, each gateway takes ~500ms-1.5s. Combining the racing + caching, overall performance on the NFTStorage gateway ends up being ~200ms.
More info:
https://nft.storage/blog/post/2022-03-08-gateway-release/
https://nft.storage/docs/concepts/gateways/#the-nftstorage-gateway
This is not helpful for the specific pathway you’re investigating (local node > public gateway), alas. But it is probably helpful in thinking about overall tradeoffs. (edited)
 -->
<!--
### considerations for apps

- Loading time can vary depending on availability.

> Lidel: where performance matters, add logic that "prewarms" HTTP CDNs/cache, and make sure etag and cache-control headers returned by http gateway are not lost. when go-ipfs 0.13 or later is used, leverage `X-Ipfs-Roots` header for even smarter HTTP cache invalidation. -->

> 

Reviewer: I think it is a good idea to admit the elephant in the room: potential abuse, and give folks some easy tips around this:

```suggestion
<!--

> Lidel2: Decrease potential load and abuse by adjusting types of hosted content
  - Consider safelisting / [denylisting](https://badbits.dwebops.pub/) specific content paths or CIDs 
  - When HTML hosting is not required (e.g. if you use IPFS for static content like pictures and videos), disable HTML rendering  by returning `Content-Disposition  attachment` HTTP header with every response. 
-->

```

Author: Absolutely!
